Cusolver: Internal review version

haxushu opened the pull request:
Internal review version

darelbeida commented on source/src_pw/hamilt_pw.cu:
```diff
@@ -1553,7 +1553,7 @@ void Hamilt_PW::add_nonlocal_pp_cuda(
             GlobalC::wf.npwx));
 	}
-	if(m == 1)
+	if(m != 1)
```
可以单独提一个PR修这个错误

haxushu replied:
https://github.com/deepmodeling/abacus-develop/issues/736

darelbeida commented on source/src_pw/energy.cpp:
```diff
@@ -244,6 +244,10 @@ void energy::print_etot(
 	}
+	else if(GlobalV::KS_SOLVER=="cusolver")
+	{
+        label = "CU";
```
缩进

haxushu replied:
和上面的缩进风格保持了一致😂

haxushu replied:
上面的缩进两种风格都有😂
<img width="563" alt="image" src="https://user-images.githubusercontent.com/25193323/158819086-1b219328-920e-4be1-aa26-d8b1c23bf234.png">

darelbeida commented on source/src_pdiag/pdiag_double.h:
```diff
@@ -56,6 +57,17 @@ class Pdiag_Double : public Pdiag_Basic
+#endif
+
+
+
```
空行数-2

haxushu replied:
done

darelbeida commented on source/src_pdiag/pdiag_double.cpp:
```diff
@@ -1272,3 +1466,108 @@ void Pdiag_Double::readin(
+	int maxncol;
+
+	MPI_Allreduce(&ncol, &maxncol, 1, MPI_INT, MPI_MAX, comm_2D);
+	// MPI_Reduce(&ncol, &maxncol, 1, MPI_LONG, MPI_MAX, 0, comm_2D);
```
注释清理一下

haxushu replied:
done

darelbeida commented on source/src_pdiag/pdiag_double.cpp:
```diff
@@ -7,7 +7,8 @@
 #include "../module_base/memory.h"
 #include "../module_base/timer.h"
-
+#include <chrono>
```
关于chrono这部分应该都是调试的，可以在适当时候清理掉

haxushu replied:
Have been removed.

darelbeida commented on source/src_pdiag/pdiag_double.cpp:
```diff
@@ -490,6 +497,8 @@ void Pdiag_Double::divide_HS_2d
 		this->nb = GlobalV::NB2D; // mohan add 2010-06-28
 	}
+
+    if (GlobalV::KS_SOLVER=="cusolver") this->nb = 1;
```
确认一下这里的缩进是因为上下文用了tab，还是你确实用了不同的空格数，如果前者可以不管，后者的话最好与上下文对齐

haxushu replied:
Have adjusted.

darelbeida commented on source/src_pdiag/pdiag_double.cpp:
```diff
@@ -586,6 +595,7 @@ void Pdiag_Double::diago_double_begin(
 				ofs<<std::endl;
 			}
 		};
+		std::cout << "Matrix size:" << GlobalV::NLOCAL << std::endl;
```
清理一下调试代码

haxushu replied:
Have been removed.

darelbeida commented on source/src_pdiag/pdiag_double.cpp:
```diff
@@ -595,6 +605,7 @@ void Pdiag_Double::diago_double_begin(
 	#endif
 
 #ifdef __MPI
+
```
清理一下无必要的改动

haxushu replied:
Have been removed.

darelbeida commented on source/src_pdiag/pdiag_double.cpp:
```diff
@@ -866,14 +877,74 @@ void Pdiag_Double::diago_double_begin(
+		ModuleBase::timer::tick("Diago_LCAO_Matrix","gath_HS");
+
+		double *htot, *stot, *vtot;
+		vtot = new double[GlobalV::NLOCAL * GlobalV::NLOCAL];
```
为什么 [L892 ](https://github.com/haxushu/abacus-develop/pull/3/files#diff-33af91a5cd00a4fb8234201f3c81e1418dd7163ac76947984c2034797908629dR892)用了vector，这里用了new，是否可以统一？

haxushu replied:
这里当时是对照着abacus之前的一个求解器写的，好处是自动析构释放空间，不会忘了delete[]。
<img width="651" alt="image" src="https://user-images.githubusercontent.com/25193323/159162047-264f4fb2-61d9-4a62-a749-76bf3fb6c6df.png">

darelbeida commented on source/src_pdiag/pdiag_double.cpp:
```diff
@@ -866,14 +877,74 @@ void Pdiag_Double::diago_double_begin(
+		double *htot, *stot, *vtot;
+		vtot = new double[GlobalV::NLOCAL * GlobalV::NLOCAL];
+		ModuleBase::Memory::record("Pdiag_Basic","vtot",GlobalV::NLOCAL*GlobalV::NLOCAL,"double");
+		if(myid==0)    // htot  NLOCAL * NLOCAL
```
用`rank`之类的词去代替`myid`

haxushu replied:
整个abacus对当前进程的命名统一为myid

darelbeida commented on source/src_pdiag/pdiag_double.cpp:
```diff
@@ -866,14 +877,74 @@ void Pdiag_Double::diago_double_begin(
+			ModuleBase::Memory::record("Pdiag_Basic","stot",GlobalV::NLOCAL*GlobalV::NLOCAL,"double");
+		}
+        
+		cuGather_double(h_mat, htot);
```
可以把`cuGather`换一个词，因为这里还没有用到cuda，会引起误解

haxushu replied:
Have changed to RootGather

darelbeida commented on source/src_pdiag/pdiag_double.cpp:
```diff
@@ -866,14 +877,74 @@ void Pdiag_Double::diago_double_begin(
+
+		ModuleBase::timer::tick("Diago_LCAO_Matrix","gath_HS");
+
+		if(myid == 0 && !diag_cusolver_gvd.init_flag){
```
`init_flag`改成`is_init`

haxushu replied:
Have changed.

darelbeida commented on source/src_pdiag/pdiag_double.cpp:
```diff
@@ -1272,3 +1466,108 @@ void Pdiag_Double::readin(
+    MPI_Status status;
+
+	if (myid == 0){
+		for (int i =0; i < GlobalV::NLOCAL; i++){
```
可以考察一下通信代价和拷贝的代价，到底是把数据先拷贝连续了，然后做一次scatter更划算？

darelbeida commented on CMakeLists.txt:
```diff
@@ -13,7 +13,8 @@ project(ABACUS
 option(ENABLE_DEEPKS "Enable DeePKS functionality" OFF)
 option(ENABLE_LIBXC "Enable LibXC functionality" OFF)
-option(USE_CUDA "Enable support to CUDA." OFF)
+option(USE_CUDA_PW "Enable support to CUDA for PW." OFF)
```
我建议这个宏不改名字，并且你用的新的宏叫`USE_CUDA_LCAO`。不改名字的原因是牺牲一些命名的精确性，换取对pw模块的不改动，并且将来解决了你遇到的那个bug之后，`USE_CUDA_LCAO`可以被`USE_CUDA`吸收

haxushu replied:
Have accepted.

darelbeida commented on CMakeLists.txt:
```diff
@@ -97,6 +109,31 @@ if(USE_CUDA)
+    -lnvToolsExt
+  )
+  set_property(TARGET ${ABACUS_BIN_NAME}
+  PROPERTY CUDA_ARCHITECTURES
```
这里方便把A100也带上吗？

haxushu replied:
我参考 https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/ 加了80 for A100

darelbeida commented on source/input.cpp:
```diff
@@ -2697,6 +2697,15 @@ void Input::Check(void)
+			{
+				
+#ifdef __MPI
+				GlobalV::ofs_warning << "Cusolver is now only available for LCAO." << std::endl;
```
全局的`cusolver`可以统一一下大小写，感觉和后面的写法不太一致，找一个统一的写法

darelbeida replied:
我不是说和`ks_solver`里面的字符串内容，我是说后面又看到了`cuSolver`写法，与这里的`Cusolver`不一致

haxushu replied:
后面都是全小写的，这里因为是一个句子，首字母大写了一下。

haxushu replied:
除了这个我后面都check统一一下

darelbeida commented on source/module_deepks/LCAO_deepks_vdelta.cpp:
```diff
@@ -107,7 +107,7 @@ void LCAO_Deepks::add_v_delta(const UnitCell_pseudo &ucell,
                             int iic;
 
-                            if(GlobalV::KS_SOLVER=="genelpa" || GlobalV::KS_SOLVER=="scalapack_gvx")  // save the matrix as column major format
+                            if(GlobalV::KS_SOLVER=="genelpa" || GlobalV::KS_SOLVER=="scalapack_gvx" || GlobalV::KS_SOLVER=="cusolver")  // save the matrix as column major format
```
这个判断后面用了非常多次，能不能新构造一个`GlobalV`的变量记录一下这个布尔值？

darelbeida replied:
变量名可能和列优先相关

haxushu replied:
abacus对genelpa和scalapack_gvx都没有这样做，感觉判断的代价不大。

darelbeida commented on source/module_cusolver/diag_cusolver.cu:
```diff
@@ -0,0 +1,245 @@
+int Diag_cuSolver_gvd::Dngvd_double(int N, int M, double *A, double *B, double *W, double *V){
+
+        copy_double(N, M, A, B);
+        buffer_double();
```
这步有办法只做第一次吗，后面只需要判断形状是否与之前一致，一致就不重复做，否则销毁`d_work`并重新做

darelbeida replied:
补充一下，这个我评论错了，需要改在`source/src_pdiag/diag_cusolver.cu`里，而不是这个为了测试而存在的文件里，需要特别注意

haxushu replied:
cusolverDnDsygvd_bufferSize不清楚内部机制，不能确定是否一致。

darelbeida commented on source/module_cusolver/diag_cusolver.cu:
```diff
@@ -0,0 +1,245 @@
+}
+int Diag_cuSolver_gvd::Dngvd_double(int N, int M, double *A, double *B, double *W, double *V){
+
+        copy_double(N, M, A, B);
```
我建议`copy`和`recopy`用上`CHECK_CUDA`宏之后，可以直接把代码写在这个函数里，加上一行注释即可，不需要单独的类方法。

darelbeida replied:
`compute`也是，核心只有一个`cusolverDnDsygvd`调用和一个sync，足够简短可以写到这个函数里面来

haxushu replied:
done

darelbeida commented on source/module_cusolver/diag_cusolver.cu:
```diff
@@ -0,0 +1,245 @@
+}
+
+
+void Diag_cuSolver_gvd::finalize(){
```
和析构函数合并起来吧，看起来并没有除了析构以外的逻辑会调用它

haxushu replied:
Have accepted.

darelbeida commented:
comment one

darelbeida reviewed (changes requested):
test

darelbeida commented on source/module_cusolver/test/cusolver_test.cpp:
```diff
@@ -0,0 +1,80 @@
+	double eig[dim];				// reference
+	
+	// construct the matrix A and B w.r.t the problem Ax = \lambda Bx
+	for (int i = 0; i < dim*dim; i++) a[i] = {1, 0};
```
这里能不能构造出一个带虚部的例子？

haxushu replied:
Hermite矩阵的特征值和广义特征值一定是实的

haxushu replied:
我再加一组

haxushu replied:
Have added to test complex_2.

darelbeida commented on tests/module_cusolver/CMakeLists.txt:
```diff
@@ -0,0 +1,126 @@
+
+# compile the test executable
+file(GLOB TEST_SRC 
+    ../../source/module_cusolver/test/cusolver_test.cpp
```
我在测试的cpp文件内以及这里都没有看到对`CUDA_LCAO`或者`MPI`宏的判断，我觉得至少应该加在某处？

haxushu replied:
考虑到单元测试的独立性，如果要分开测，可以在file(GLOB TEST_SRC单列cusolver_test.cpp或mpi.cpp

caic99 committed:
Merge pull request #817 from dyzheng/new

fix : UT for diago_cg.cpp and diago_david.cpp should not depend on __CUDA and __ROCM

dyzheng committed:
fix : support both element magnetization and atomic magnetization for initial

dyzheng committed:
Merge pull request #118 from PeizeLin/develop

1. change HONG in Makefile

dyzheng committed:
add 304_NO_GO_AF_atommag case for testing initial atomic magnetization

dyzheng committed:
fix bug: initial magnetization method (#119)

* fix : when input structure doesn't have most symmetry rotation matrix,
ABACUS will print which structure is better and then using the input structure for calculating.

* perf : add warning output for symmetry searching part; modified tests/integral/207_NO_DP_OD/result.ref to correct version.

* fix : UT for diago_cg.cpp and diago_david.cpp should not depend on __CUDA and __ROCM

* fix : support both element magnetization and atomic magnetization for initial

* add 304_NO_GO_AF_atommag case for testing initial atomic magnetization

dyzheng committed:
Merge pull request #818 from dyzheng/new

fix : support both element magnetization and atomic magnetization for initial setting in STRU file

dyzheng committed:
merge deepmodeling/abacus-develop/develop branch with develop branch (#120)

* fix : when input structure doesn't have most symmetry rotation matrix,
ABACUS will print which structure is better and then using the input structure for calculating.

* perf : add warning output for symmetry searching part; modified tests/integral/207_NO_DP_OD/result.ref to correct version.

* fix : UT for diago_cg.cpp and diago_david.cpp should not depend on __CUDA and __ROCM

* fix : support both element magnetization and atomic magnetization for initial

* add 304_NO_GO_AF_atommag case for testing initial atomic magnetization

Co-authored-by: Chun Cai <amoycaic@gmail.com>

dyzheng committed:
Merge pull request #819 from abacusmodeling/develop

synchronization between abacusmodeling and deepmodeling abacus-develop/develop branch

dyzheng committed:
Merge pull request #121 from deepmodeling/develop

merge

pxlxingliang committed:
fix bug: the opended file 'data-H' is not closed
range: source/src_pw/test/diago_david_test.cpp

PeizeLin committed:
1. update Makefile.system and Makefile.vars

pxlxingliang committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into develop

dyzheng committed:
Merge pull request #122 from PeizeLin/develop

1. update Makefile.system and Makefile.vars

dyzheng committed:
Fix : occupied band should be number of electrons instead of half

dyzheng committed:
Merge branch 'develop' of github.com:deepmodeling/abacus-develop into new

maki49 committed:
fix uninited rcutmax

Commit:
deepks : allow reading embedding model

Commit:
Merge branch 'develop' of github.com:deepmodeling/abacus-develop into develop

dyzheng committed:
Merge pull request #821 from maki49/develop

fix uninited rcutmax

pxlxingliang committed:
Merge branch 'develop' of github.com:deepmodeling/abacus-develop into develop

caic99 committed:
Merge pull request #820 from pxlxingliang/develop

fix bug: the opended file 'data-H' is not closed

YuLiu98 committed:
rewrite function of MD_dump

dyzheng committed:
Merge pull request #822 from wenfei-li/develop

deepks : allow reading embedding model

Commit:
deepks : attempt to fix search_radius error in 'unit' test

Commit:
Merge branch 'develop' of github.com:deepmodeling/abacus-develop into develop

Commit:
deepks : enable output log if ut failed

dyzheng committed:
Merge branch 'develop' of github.com:deepmodeling/abacus-develop into new

dyzheng committed:
Fix: removed 2.1.0 version manual documentation in doc directory

maki49 committed:
fix a delete in parallel_orbitals

caic99 committed:
Merge pull request #827 from maki49/patch-1

fix a delete in parallel_orbitals

caic99 committed:
Merge pull request #825 from wenfei-li/develop

deepks : attempt to fix search_radius error in 'unit' test

dyzheng committed:
Merge pull request #824 from YuLiu98/develop

rewrite MD_dump when step=0

Qx80610702 committed:
Adjust bands output for DMFT

Qx80610702 committed:
Adjust bands output for DFT+DMFT

caic99 committed:
Fix: add `.gitignore` items for generated files in `tests/deepks/*`.

caic99 committed:
Build: fix sanitizer test logic.
Using sanitizer is most likely produce failure. In this case,
the result should also be published.

Commit:
gint : removed obsolete codes

dyzheng committed:
Merge pull request #829 from caic99/develop

Fix: add `.gitignore` items for generated files in `tests/deepks/*`.

caic99 committed:
Modify UT branches.

dyzheng committed:
Merge pull request #831 from wenfei-li/develop

gint : removed obsolete codes

dyzheng committed:
Merge pull request #832 from caic99/develop

Build: fix sanitizer test logic.

Commit:
xc_functional : explicitely set use_libxc = true again for some cases

dyzheng committed:
Merge pull request #123 from Qx80610702/develop

Adjust bands output for DFT+DMFT

PeizeLin committed:
1. fix bug in LOOP_elec::solver()
2. fix bug in XC_Functional::set_xc_type()

dyzheng committed:
Merge pull request #124 from PeizeLin/develop

1. fix bug for 'dft_functional = hse' in LOOP_elec::solver()

caic99 committed:
Test: change testing target to merged reference

dyzheng committed:
Merge pull request #834 from abacusmodeling/develop

Merge : 1, Makefile fix, 2, DMFT output, 3, hse bug fix

dyzheng committed:
Merge pull request #835 from caic99/develop

Test: update UT branches

dyzheng committed:
Merge pull request #826 from dyzheng/new

Fix: deleted useless file, output right occupied band for soc case

caic99 committed:
Merge branch 'develop' of
https://github.com/deepmodeling/abacus-develop into develop

caic99 committed:
Build: fix sanitizer report path

pxlxingliang committed:
test: add the unit test of diagonalization by ELPA
range: source/src_pdiag

caic99 committed:
Merge pull request #838 from caic99/develop

Test: change testing target to merged reference

caic99 committed:
Build: Add `libnpy`, `googletest` in docker env.

dyzheng committed:
Refactor: move Global.h dependency from header to cpp files, for UT coding, only exx unfinished

caic99 committed:
Fix: potential array access overflow in gamma vl

Qianruipku committed:
Update esolver.h

add virtual deconstructor for esolver.h

caic99 committed:
Update issue templates

- Revise "Bug Report"
- Add "Help Wanted"
- Select appropriate tag for each issue template

dyzheng committed:
Merge pull request #843 from Qianruipku/hotfix

Update esolver.h

dyzheng committed:
Merge pull request #844 from deepmodeling/caic99-patch-1

Update issue templates

dyzheng committed:
Merge pull request #842 from dyzheng/new

Refactor: move Global.h dependency from header to cpp files,

caic99 committed:
Merge pull request #840 from caic99/develop

Build: Add `libnpy`, `googletest` in docker env.

caic99 committed:
Build: fix file path of `libnpy`.

pxlxingliang committed:
- move function read_hs() to diago_elpa_utils.h
- rename H-KPonts.dat to be H-KPoints.dat
- delete some redundant information output

dyzheng committed:
Merge pull request #845 from caic99/develop

Fix: potential array access overflow in gamma vl

caic99 committed:
Build: fix static analysis workflow logic.

dyzheng committed:
Merge pull request #847 from caic99/develop

Build: fix static analysis workflow logic.

caic99 committed:
Test: Fix new-delete mismatch

caic99 committed:
Fix: array access out of bound

caic99 committed:
Fix: misuse of '=' in assertion

caic99 committed:
Merge pull request #850 from caic99/develop

Fix: misuse of '=' in assertion

caic99 committed:
Fix: variable not initialized in for-loop

caic99 committed:
Merge pull request #851 from caic99/develop

Fix: variable not initialized in for-loop

caic99 committed:
Build: update testing strategy
Before, we are using `pull_request_target` to access AWS token
to start a self-hosted runner. Since we are not using it, changing to
`pull_request` should be more secure, for this trigger can not access
repositry token.

caic99 committed:
Merge pull request #852 from caic99/develop

Build: update testing strategy

caic99 committed:
Test: rename CI step
Separate Build and Test to two steps.

caic99 committed:
Merge pull request #853 from caic99/develop

Test: rename CI step

caic99 committed:
CI: Enable static code analysis.

caic99 committed:
Merge pull request #854 from caic99/develop

CI: Enable static code analysis.

caic99 committed:
Doc: Format markdown files.

dyzheng committed:
Merge: HSolver to develop, diagonalization for PW and LCAO has finished. (#848)

* refactor : add base class in module_hsolver, module_hamilt, module_psi, module_elecstate
compiling for diagocg.cpp, hamiltpw.cpp, elecstatepw.cpp, psi.cpp is passed.

* fix psi.h compiling error

* refactor : add DiagoCG class in HSolver module, modified Psi, Hamilt, ElecState, Charge, PW_Basis classes

* fix : delete useless variable Bec and fix compiling bug

* Refactor: added diagodavid.cpp and iterdiagcon.cpp to module_hsolver

* Fix: bug in psi.h

* Refactor: move diagH_subspace and diagH_LAPACK to IterDiagControl, lcao_in_pw with exx is not supported yet

* Refactor: finished diag() function in diagocg.cpp and diagodavid.cpp

* Fix: bug in hsolverpw.cpp, psi not update for each k point

* Fix: compiling error

* Fix: compiler error for mutable pointer in const function

* Refactor: temporary deleted interface init and update in HSolver

* Refactor: changed use of NBANDS to psi.get_nbands()

* Refactor: added HSolverLCAO class

* Refactor: added example of template class with virtual function

* Fix: matrix() in HamiltLCAO

* Refactor: added fixed realspace matrix to HamiltLCAO

* Fix: modified Psi class constructor for UT, some update for hamiltlcao

* Fix: removed PW_Basis dependency in DiagoCG and DiagoDavid

* Fix: template function only can be defined explicitly in cpp files, fixed psi.h

* Fix: int* ngk in psi.h should be const type

* UT of diagocg (not successful)

* Fix: bug in diagodavid.h

* Fix: bug in psi.h

* Fix: template function in hamiltlcao.cpp should be modified

* UT of cg in HSolver (successful but fragile)

* add unit test of DAVIDSON diagonalization for HSolver

* Refactor: T, NL, S matrix calculation for HamiltLCAO

* fix the bug of getpsi() in diao_mock.h

* Fix: allocate matrix with correct size in HamiltLCAO class

* Fix: compiling error in HamiltLCAO class

* Refactor: move diag_scalapack_gvx to module_hsolver/diagosca

* Refactor: added DiagoElpa and HSolverLCAO to HSolver module

* Fix: conflict with CUDA has fixed for merge into develop branch

* Fix: add template<> for .cpp files

* Fix: change namespace name for hsolver, hamilt, elecstate, psi

* Fix: bug of psi and added template class specify in cpp

* Refactor: modified name of files in new modules

* Refactor: format the code

Co-authored-by: root <hongriTianqi>
Co-authored-by: xingliang <pxlppq007@163.com>
Co-authored-by: pxlxingliang <91927439+pxlxingliang@users.noreply.github.com>

caic99 committed:
Test: fix math_bspline_test.cpp

Delete duplicated tests.

caic99 committed:
doc: move `developers.md` to `CONTRIBUTING.md`.

Following naming conventions in https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/creating-a-default-community-health-file .

dyzheng committed:
Merge pull request #855 from caic99/develop

Doc: Format markdown files.

pxlxingliang committed:
close the print out of redundant information

pxlxingliang committed:
modify S matrix not to be a unitary matrix

caic99 committed:
Merge pull request #856 from caic99/develop

docs: move `developers.md` to `CONTRIBUTING.md`.

YuLiu98 committed:
fix: update the default value of md_tfreq

YuLiu98 committed:
Merge branch 'develop' of https://github.com/deepmodeling/abacus-develop into develop

YuLiu98 committed:
refactor(md): update output format of MD.

add scf iteration title for pw md, update output format of md

caic99 committed:
docs: fix path link to `docs/`

pxlxingliang committed:
- test elpa by real Si system produced H/S matrix
- modify the parallel test to be only run one time

pxlxingliang committed:
delete the install of unnecessary file diago_parallel_test.sh

dyzheng committed:
Merge pull request #857 from YuLiu98/develop

update the default value of md_tfreq

dyzheng committed:
Merge pull request #858 from caic99/develop

docs: fix path link to `docs/`

dyzheng committed:
Merge pull request #839 from pxlxingliang/develop

test: add the unit test of parallel diagonalization by ELPA

Commit:
Merge branch 'develop' of github.com:deepmodeling/abacus-develop into HEAD

Commit:
modify comment, doumentation and HSE example INPUT

Commit:
fix a type in documentation

caic99 committed:
Merge pull request #859 from wenfei-li/develop

modify documentation, comment, and HSE input

Commit:
UT of Charge::sum_band()

pxlxingliang committed:
test: add the unit test of ELPA/SCALAPACK diagonalization

dyzheng committed:
Merge pull request #860 from hongriTianqi/develop

test(pw): add UT of `Charge::sum_band()`

dyzheng committed:
Merge pull request #861 from pxlxingliang/develop

test(LCAO): add the UT of ELPA/SCALAPACK diagonalization

sunliang98 committed:
<fix>
Add a parameter "ifwarn" to the SCAN_BEGIN and SCAN_END in global_function.h to switch the warning.
Fix a output bug in Pseudopot_upf::read_pseudo_upf201, which may out invalid warning. For example, if the key word in pseudopotential is <PP_R>, ABACUS will look for <PP_R first and output the warning that the <PP_R> isn't found, next it will find <PP_R>, but <PP_R and <PP_R> are equivalent, so the warning is invalid.
As a result, I turn off the warning in the functions that look for <PP_R, <PP_RAB, <PP_DIJ, <PP_RHOATOM.
Besides, many pseudopotentials do not have block <PP_SPIN_ORB>, the warning seems needless, so I turn off the warning of looking for <PP_SPIN_ORB> and </PP_SPIN_ORB>.
The warning of </UPF> is turned off for the same reason.
The problems of <PP_SPIN_ORB> and </UPF> are mentioned in issue#128 in abacusmodeling/abacus-develop.

ddhhss committed:
Feature: Add module surchem. (#846)

* add H_Hartree files.

* fix bug of ngmc/nrxx ..

* fix bug, CG test PASS.

* fix eb_k

* print square of residue

* change dim of epsilon from nrxx*2 to nrxx*1..

* add get_atom_num and Vcav

* add test for rho

* fix test

* debug of gauss charge.

* add Vcav and Vel

* delete malloced vectors, fix bug of segfault in function createcavity.

* fix bug: no match for 'operator*' (operand types are 'std::complex<double>' and 'int')

* remove redundant annotations.

* add v_cav and v_el.

* Use laplace(vloc) to N_nuclear, delete redundant comments.

* add test for N & n

* fix N nuclear / 4pi

* Add calculation of Acav, Ael, and Hamitonian.

* add globalV

* change cmakelists

* update input file

* add module surchem into link libraries of deepks_test.

Co-authored-by: sunml99 <sunmenglin@stu.pku.edu.cn>

sunliang98 committed:
Add annotation to the parameter ifwarn in SCAN_BEGIN and SCAN_END

Commit:
test[pw]: cg of HSolver with larger H matrices

Commit:
modify diago_cg_test.cpp

dyzheng committed:
Fix: 1. only init wfc once for MD and relax; 2. reset PW_DIAG_THR every Ionic step. 3. 17 test cases reference updated

dyzheng committed:
Merge branch 'develop' of github.com:deepmodeling/abacus-develop into develop

dyzheng committed:
Merge pull request #863 from hongriTianqi/develop

test(pw): cg of HSolver with larger H matrices

dyzheng committed:
Merge pull request #862 from sunliang98/debug

Fix: remove invalid warning in `Pseudopot_upf::read_pseudo_upf201`.

maki49 committed:
pull develop and solve conflicts

maki49 committed:
modify wfc allocation to pass integer

dyzheng committed:
Merge pull request #864 from dyzheng/develop

Fix: 1. only init wfc once for MD and relax; 2. reset PW_DIAG_THR every Ionic step. 3. 17 test cases reference updated

dyzheng committed:
Fix: 2D-block allolation error

dyzheng committed:
Fix : don't print cube file if out_chg is 0

dyzheng committed:
Merge pull request #868 from dyzheng/develop

Fix: 1, 2D-block allocation error. 2, not print charge density cube file if out_chg is 0

maki49 committed:
Merge branch 'develop' of https://github.com/deepmodeling/abacus-develop into LCAO-refactor

ddhhss committed:
Fix：Added support for makefile to module_surchem. (#871)

dyzheng committed:
Refactor: add elecstate.cpp and elecstate_pw.cpp, HSolverPW is finished

dyzheng committed:
Refactor: move ebandK() to elecstate.h

dyzheng committed:
Refactor: add no_subspace judgement in DiagoCG

pxlxingliang committed:
test(hsolver):add the parallel test of CG and DAVIDSON diagonalization

dyzheng committed:
Merge pull request #865 from pxlxingliang/develop

test(hsolver):add the parallel test of CG and DAVIDSON diagonalization

Qianruipku committed:
< refactor >
   Write Run function for ESovler_KS.
   Prepare SCF program for all ESolver_KS-derived Class
   All tests passed.

pxlxingliang committed:
fix(diago):the dimension of eigenvalue array in scalapack interface

Qianruipku committed:
<fix>
	fix a bug that make cannot compile ABACUS

dyzheng committed:
Refactor: added cal_dm in ElecStateLCAO, psi_k_2d->dm_k_2d

dyzheng committed:
Merge branch 'develop' of https://gitee.com/deepmodeling/abacus-develop into HSolver

pxlxingliang committed:
Merge pull request #873 from pxlxingliang/develop

fix(hsolver): correct the dimension of input eigenvalue array in scalapack interface

Qianruipku committed:
<Refactor>
	Rewrite ESolver_KS_PW with the SCF frame in ESolver_KS
	The order of functions are not changed, and in other words, we do not optimize the position of some functions yet.
	Some changes we make:
		1. Here we do not consider the colorful screen print with GlobalV::COLOUR.
		2. We do not print CGITER in the screen and write it in the running.log instead.
		3. We do not move cal_exx() to ESolver_KS_PW
			if( Exx_Global::Hybrid_Type::HF   == GlobalC::exx_lcao.info.hybrid_type ||
			Exx_Global::Hybrid_Type::PBE0 == GlobalC::exx_lcao.info.hybrid_type ||
			Exx_Global::Hybrid_Type::HSE  == GlobalC::exx_lcao.info.hybrid_type )
        	{
            	if( !GlobalC::exx_global.info.separate_loop )
            	{
                	GlobalC::exx_lip.cal_exx();
            	}
            	break;
			}
		which is in Electrons class before. I doubt if it is useful.

Qianruipku committed:
Update Makefile.vars

Qianruipku committed:
modify the screen output

pxlxingliang committed:
fix(hsolver):initialize static variable `is_already_decomposed`

Qianruipku committed:
<fix>
	Now out_chg is the parameter deciding if to output charge and out_freq_elec and out_freq_ion decide the frequency.

dyzheng committed:
Merge pull request #876 from pxlxingliang/develop

fix(hsolver):initialize static variable `is_already_decomposed`

pxlxingliang committed:
fix(hsolver):correct the input of set_elpahandle of
complex<double> case

Qianruipku committed:
change afteriter/beforeiter to afterscf/beforescf

Qianruipku committed:
Change beforeiter/afteriter to beforescf/afterscf

Qianruipku committed:
Merge branch 'SDFT' of https://github.com/Qianruipku/abacus-develop into SDFT

dyzheng committed:
Merge pull request #877 from pxlxingliang/develop

fix(hsolver):correct the input of set_elpahandle in complex<double> case

Qianruipku committed:
Merge pull request #875 from Qianruipku/SDFT

<refactor>

Commit:
gint : 1. removed more obsolete codes;
2. add new implementation of gamma force

pxlxingliang committed:
test(hsolver): add the test of ELPA/SCALAPACK

Commit:
test[LCAO]: UT of functions in gint_gamma_rho.cpp

Commit:
modify source/src_lcao/CMakeLists.txt

Commit:
remove .dat in LOWF_GAMMA_S1.dat because gitignore

Commit:
modify gamma_rho_mock.h

dyzheng committed:
Merge pull request #879 from pxlxingliang/hsolver

test(hsolver): add the test of ELPA/SCALAPACK

dyzheng committed:
Merge pull request #880 from wenfei-li/develop

gint : 1. removed more obsolete codes; 2. add new implementation of gamma force

maki49 committed:
Merge branch 'develop' of https://github.com/deepmodeling/abacus-develop into LCAO-refactor

maki49 committed:
Merge branch 'SDFT' of https://github.com/deepmodeling/abacus-develop into LCAO-refactor

Commit:
modify default pseudo_type in global_variable

Commit:
modify comments in global_variable.cpp

ouqi0711 committed:
implemented DeePKS orbital(bandgap) label for multi-k case

ouqi0711 committed:
Merge https://github.com/ouqi0711/abacus-develop into develop

Qianruipku committed:
create ESolver_SDFT_PW

Qianruipku committed:
change all self_consistent(istep-1) to p_esolver->Run(istep-1)

Qianruipku committed:
fix a bug that can lead to segmental fault in 817_NO_LT_fcc

dyzheng committed:
Refactor: finished HSolverLCAO class, gamma_only and multi-k lines

dyzheng committed:
Merge branch 'HSolver' of github.com:deepmodeling/abacus-develop into HSolver

dyzheng committed:
Refactor: format the code

mohanchen committed:
Merge pull request #887 from Qianruipku/hotfix

fix a bug that can lead to segment fault in 817_NO_LT_fcc

mohanchen committed:
Merge pull request #888 from deepmodeling/HSolver

Refactor: added ElecStatePW and ElecStateLCAO for finishing HSovlerPW and HSolverLCAO

Qianruipku committed:
Merge branch 'develop' of https://github.com/deepmodeling/abacus-develop into SDFT

Qianruipku committed:
delete repetition in Makefile.Objects

maki49 committed:
remove redundant module_surchem in Makefile

Qianruipku committed:
fix a bug that makes tests 108 & 109 fail.

Qianruipku committed:
Merge branch 'SDFT' of https://github.com/deepmodeling/abacus-develop into SDFT

Qianruipku committed:
delete Electrons::self_consistent

Qianruipku committed:
fix a bug, change Run(istep-1) to Run(istep)

Qianruipku committed:
fix a bug cuda cannot be compiled

maki49 committed:
move orb_con, RA, LOC, LOWF, LM, UHM into ESolver_KS_LCAO; add cal_DOS to ESolver; remove final_scf

maki49 committed:
pull develop and solve conflicts

mohanchen committed:
Merge pull request #890 from deepmodeling/SDFT

Merge SDFT branch to develop branch.

maki49 committed:
pull LCAO-refactor and solve conflicts

Qianruipku committed:
add out_freq_elec and out_freq_ion into input-main.md

maki49 committed:
fix nbands=0 bug

maki49 committed:
delete LOOP_elec and modify beforescf

maki49 committed:
rename cal_DOS() as postprocess()

Qianruipku committed:
add const parameter

Commit:
add non-MPI code in WF_Local::distri_lowf_new()

mohanchen committed:
Merge pull request #892 from Qianruipku/hotfix

refactor: modify input-main.md and add const for parameters

Qianruipku committed:
for lcao and lcao_in_pw diag_ethr = 0.0

mohanchen committed:
Merge pull request #893 from Qianruipku/SDFT

only for tests

wenfei-li committed:
Merge pull request #884 from ouqi0711/develop

DeePKS orbital (bandgap) label for multi-k case

Commit:
gint : removed obsolete code for k point force

Commit:
gint : cal_pvpR_reduced, change t1<=t2 to iat1<=iat2

Commit:
gint : removed more obsolete code from multi-k force

hongriTianqi committed:
Merge pull request #881 from hongriTianqi/develop

test(LCAO): UT of functions in gint_gamma_rho.cpp

dyzheng committed:
Merge pull request #894 from wenfei-li/develop

Gint : 1. remove obsolete code; 2. t1<=t2 to iat1<=iat2 in cal_pvpR_reduced

Commit:
gint_k : removed obsolete code evaluate_pDMp

maki49 committed:
pull SDFT and solve conflicts

maki49 committed:
pull develop

maki49 committed:
add istep parameter to beforescf

maki49 committed:
add OBJS_SURCHEM in Makefile

pxlxingliang committed:
test(integrate): add the test of kpar

pxlxingliang committed:
modify the jd

pxlxingliang committed:
Merge pull request #899 from pxlxingliang/develop

test(integrate): add the test of kpar

pxlxingliang committed:
test(performance): add performance test exampels

Commit:
gint : implement new method for gamma forcde & stress;
also fix a bug for old method

maki49 committed:
rearrange LCAO workflow into ESolver_KS_LCAO

Commit:
gint : unify cal_psir a little bit

maki49 committed:
fix an istep-bug

Commit:
gint : removed unnecessary code

maki49 committed:
fix a build-error in test

maki49 committed:
fix compile error

maki49 committed:
Merge branch 'develop' of https://github.com/deepmodeling/abacus-develop into LCAO-refactor

Commit:
gint : removed grid_base_beta

Commit:
gint : unify cal_psi in multi-k rho

Commit:
gint : unify a little bit gint_k_vl

Commit:
Merge branch 'develop' of github.com:deepmodeling/abacus-develop into develop

Commit:
gint : update cmakefile for src_lcao test

Commit:
gint : update gamma_rho_mock.h

dyzheng committed:
Merge pull request #900 from pxlxingliang/develop

test(performance): add performance test exampels

Commit:
test(mpi): UT of functions in parallel_common.cpp

Commit:
modify parallel_common_test.cpp

Commit:
modify typo in parallel_common_test.cpp

hongriTianqi committed:
Merge pull request #905 from hongriTianqi/develop

test(mpi): UT of functions in parallel_common.cpp

ouqi0711 committed:
added test for deepks multi-k bandgap label

hongriTianqi committed:
test(mpi): UT of 2 functions in parallel_global.cpp

ouqi0711 committed:
Merge https://github.com/ouqi0711/abacus-develop into develop

Commit:
gint : update makefile

hongriTianqi committed:
Merge pull request #906 from hongriTianqi/develop

test(Parallel): add UT of `split_diag_world` and `split_grid_world`

Commit:
Merge branch 'develop' of github.com:deepmodeling/abacus-develop into develop

wenfei-li committed:
Merge pull request #907 from ouqi0711/develop

added test case for DeePKS multi-k band gap label

darelbeida committed:
implemented a new ks_solver called "cusolver" in LCAO

wenfei-li committed:
Merge pull request #897 from wenfei-li/develop

gint : removed obsolete code; implement new gamma force & stress

maki49 committed:
revert private

maki49 committed:
move LOOP_elec.cpp to module_esolver/esolver_ks_lcao_elec.cpp

maki49 committed:
add a bool have_force in ESolver_LCAO

maki49 committed:
reformat some files

maki49 committed:
reformat all files in module_esolver

maki49 committed:
reformat some files

maki49 committed:
add comments for public variables

maki49 committed:
pull develop

maki49 committed:
fix: set have_force=false after every md step

sunliang98 committed:
<fix>
1. Fix the bug that multiple line heading of blocks <PP_R>, <PP_RAB>, <PP_DIJ>, and <PP_RHOATOM> in "PROJECT-Eu.UPF-1.txt" cannot be recognized by Pseudopot_upf::read_pseudo_upf201.
2. Release the restriction that off-diagonal elements of DIJ (Pseudopot_upf::dion) must be zero.

Commit:
gint : fix bug in gamma force

wenfei-li committed:
Merge pull request #911 from wenfei-li/develop

Fix: not using `DM[is]` when no atom found on grid

sunliang98 committed:
Merge branch 'develop' of https://github.com/deepmodeling/abacus-develop into debug

dyzheng committed:
Merge pull request #910 from sunliang98/debug

fix: Add the support to the multiple line heading in pseudopotential.

maki49 committed:
Merge branch 'develop' of https://github.com/deepmodeling/abacus-develop into LCAO-refactor

hongriTianqi committed:
test(parallel): UT of methods in parallel_kpoints

hongriTianqi committed:
modify default number of processes to 4

hongriTianqi committed:
Merge pull request #912 from hongriTianqi/develop

test(parallel): UT of methods in parallel_kpoints

maki49 committed:
add new deepks-bandgap-code into esolver

dyzheng committed:
Merge pull request #909 from maki49/LCAO-refactor

Merge LCAO-refactor into develop
